transport/httptransport/http_transport.go

```diff
@@ -0,0 +1,235 @@
+		}
+		return t, nil
+	}
+	// start executing the transfer
+	t.wg.Add(1)
+	go func() {
+		defer t.wg.Done()
+		defer close(t.eventCh)
+
+		if err := t.execute(tctx); err != nil {
+			if err := t.emitEvent(tctx, types.TransportEvent{
+				Error: err,
+			}, dealInfo.DealUuid); err != nil {
+				log.Errorw("failed to publish transport error", "id", t.dealInfo.DealUuid, "err", err)
+			}
+		}
+	}()
+	return t, nil
+}
+
+type transfer struct {
+	closeOnce sync.Once
+	cancel    context.CancelFunc
+
+	eventCh chan types.TransportEvent
+
+	tInfo    *types.HttpRequest
+	dealInfo *types.TransportDealInfo
+	wg       sync.WaitGroup
+
+	nBytesReceived int64
+}
+
+func (t *transfer) emitEvent(ctx context.Context, evt types.TransportEvent, id uuid.UUID) error {
+	select {
+	case t.eventCh <- evt:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		return fmt.Errorf("dropping event %+v as channel is full for deal id %s", evt, id)
+	}
+}
+
+func (t *transfer) execute(ctx context.Context) error {
+	// construct request
+	req, err := http.NewRequest("GET", t.tInfo.URL, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create http req: %w", err)
+	}
+
+	// add range req to start reading from the last byte we have in the output file
+	req.Header.Set("Range", fmt.Sprintf("bytes=%d-", t.nBytesReceived))
+	// init the request with the transfer context
+	req = req.WithContext(ctx)
+	// open output file in append-only mode for writing
+	of, err := os.OpenFile(t.dealInfo.OutputFile, os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open output file: %w", err)
+	}
+	defer of.Close()
+
+	// start the http transfer
+	remaining := t.dealInfo.DealSize - t.nBytesReceived
+	if err := t.doHttp(ctx, req, of, remaining); err != nil {
+		// do not resume transfer if context has been cancelled
+		if ctx.Err() == context.Canceled || ctx.Err() == context.DeadlineExceeded {
+			return fmt.Errorf("transfer context err: %w", ctx.Err())
+		}
+
+		//
+		// TODO: resumption
+		return fmt.Errorf("failed to execute http transfer: %w", err)
+	}
+	// --- http request finished successfully. see if we got the number of bytes we expected.
+
+	// if the number of bytes we've received is not the same as the deal size, we have a failure.
+	if t.nBytesReceived != t.dealInfo.DealSize {
+		return fmt.Errorf("mismatch in dealSize vs received bytes, dealSize=%d, received=%d", t.dealInfo.DealSize, t.nBytesReceived)
+	}
+
+	// otherwise we're good ! notify the subscriber.
+	if err := t.emitEvent(ctx, types.TransportEvent{
+		NBytesReceived: t.nBytesReceived,
+	}, t.dealInfo.DealUuid); err != nil {
+		log.Errorw("failed to publish transport event", "id", t.dealInfo.DealUuid, "err", err)
+	}
+
+	return nil
+}
+
+func (t *transfer) doHttp(ctx context.Context, req *http.Request, dst io.Writer, toRead int64) error {
+	// send http request and validate response
+	resp, err := http.DefaultClient.Do(req)
```

Comment:
I wonder if there's some way to pass a context here 🤔 
If it takes a long time to connect to the HTTP server (eg because it's behind a firewall and hangs forever) it seems like there's no way to cancel the request.

Reply:
We already do that in:

```
 // init the request with the transfer context
	req = req.WithContext(ctx)
```

Reply:
Ah ok makes sense 👍